src/scalarstats.jl

```diff
@@ -197,75 +194,117 @@ nquantile(v::AbstractArray{<:Real}, n::Integer) = quantile(v, (0:n)/n)
 """
     variation(x, m=mean(x))
 
-Return the coefficient of variation of an array `x`, optionally specifying
+Return the coefficient of variation of collection `x`, optionally specifying
 a precomputed mean `m`. The coefficient of variation is the ratio of the
 standard deviation to the mean.
 """
-variation(x::AbstractArray{<:Real}, m::Real) = stdm(x, m) / m
-variation(x::AbstractArray{<:Real}) = variation(x, mean(x))
+variation(x, m) = stdm(x, m) / m
+variation(x) = ((m, s) = mean_and_std(x); s/m)
+
+# Standard error of the mean: std / sqrt(len)
+# Code taken from var in the Statistics stdlib module
+
+# faster computation of real(conj(x)*y)
+realXcY(x::Real, y::Real) = x*y
+realXcY(x::Complex, y::Complex) = real(x)*real(y) + imag(x)*imag(y)
 
-# Standard error of the mean: std(a) / sqrt(len)
 """
-    sem(a)
+    sem(x)
 
-Return the standard error of the mean of `a`, i.e. `sqrt(var(a) / length(a))`.
+Return the standard error of the mean of collection `x`,
+i.e. `sqrt(var(x, corrected=true) / length(x))`.
 """
-sem(a::AbstractArray{<:Real}) = sqrt(var(a) / length(a))
+function sem(x)
+    y = iterate(x)
+    if y === nothing
+        T = eltype(x)
+        return oftype(sqrt((abs2(zero(T)) + abs2(zero(T)))/2), NaN)
+    end
+    count = 1
+    value, state = y
+    y = iterate(x, state)
+    # Use Welford algorithm as seen in (among other places)
+    # Knuth's TAOCP, Vol 2, page 232, 3rd edition.
+    M = value / 1
+    S = real(zero(M))
+    while y !== nothing
+        value, state = y
+        y = iterate(x, state)
+        count += 1
+        new_M = M + (value - M) / count
+        S = S + realXcY(value - M, value - new_M)
+        M = new_M
+    end
+    var = S / (count - 1)
+    return sqrt(var/count)
+end
 
 # Median absolute deviation
 """
-    mad(v; center=median(v), normalize=true)
+    mad(x; center=median(x), normalize=true)
 
-Compute the median absolute deviation (MAD) of `v` around `center`
+Compute the median absolute deviation (MAD) of collection `x` around `center`
 (by default, around the median).
 
 If `normalize` is set to `true`, the MAD is multiplied by
 `1 / quantile(Normal(), 3/4) ≈ 1.4826`, in order to obtain a consistent estimator
 of the standard deviation under the assumption that the data is normally distributed.
 """
-function mad(v::AbstractArray{T};
-             center::Union{Real,Nothing}=nothing,
-             normalize::Union{Bool, Nothing}=nothing) where T<:Real
-    isempty(v) && throw(ArgumentError("mad is not defined for empty arrays"))
-
-    S = promote_type(T, typeof(middle(first(v))))
-    v2 = LinearAlgebra.copy_oftype(v, S)
+function mad(x; center=nothing, normalize::Union{Bool, Nothing}=nothing)
+    isempty(x) && throw(ArgumentError("mad is not defined for empty collections"))
+
+    if x isa AbstractArray
+        T = eltype(x)
+        S = promote_type(T, typeof(middle(zero(T))))
+        x2 = LinearAlgebra.copy_oftype(x, S)
+    else # Use type of first element
```

Comment:
Using the first element sounds a bit arbitrary. A possibly better alternative would be to progressively widen the element type as we encounter new types, but that potentially requires allocating several copies.

Reply:
This would be fine to progressively widen, since it is very rare because most containers are homogenious.

Reply:
Actually it sounds simpler to make a copy when the element type is not concrete. That way we choose the correct element type. Anyway performance doesn't matter a lot in that case and a copy is cheap compared with computing the median.